lib/jekyll-remote-theme/downloader.rb

```diff
@@ -0,0 +1,90 @@
+# frozen_string_literal: true
+
+module Jekyll
+  module RemoteTheme
+    class Downloader
+      include RemoteTheme::Executor
+
+      HOST = "https://codeload.github.com".freeze
+      PROJECT_URL = "https://github.com/benbalter/jekyll-remote-theme".freeze
+      USER_AGENT  = "Jekyll Remote Theme/#{VERSION} (+#{PROJECT_URL})".freeze
+      TEMP_PREFIX = "jekyll-remote-theme-".freeze
+
+      attr_reader :theme
+      private :theme
+
+      def initialize(theme)
+        @theme = theme
+      end
+
+      def run
+        if downloaded?
+          Jekyll.logger.debug LOG_KEY, "Using existing #{theme.name_with_owner}"
+          return
+        end
+
+        download
+        unzip
+        set_theme_root
+
+        @downloaded = true
+      end
+
+      def downloaded?
+        @downloaded ||= theme_dir_exists? && !theme_dir_empty?
+      end
+
+      def temp_dir
+        @temp_dir ||= File.realpath Dir.mktmpdir(TEMP_PREFIX)
+      end
+
+      private
+
+      def zip_file
+        @zip_file ||= Tempfile.new([TEMP_PREFIX, ".zip"])
+      end
+
+      def download
+        Jekyll.logger.debug LOG_KEY, "Downloading #{zip_url} to #{zip_file.path}"
+        cmd = [
+          *timeout_command, "curl", "--url", zip_url, "--output", zip_file.path,
+          "--user-agent", USER_AGENT, "--fail", "--silent", "--show-error",
+        ]
+        run_command(*cmd)
```

Comment:
Maybe for a later iteration, but retries here might be nice. A network blip could cause downloads to fail so giving this a `rescue ExecutionError` with some recursion might be helpful to prevent these kinds of blips.

```ruby
def download(tries: 3)
  run_command(*cmd)
rescue ExecutionError => e
  raise e if tries <= 0
  download(tries: tries - 1)
end
```

❓

Reply:
Agree, but hoping to follow up in a subsequent pass with #2 and #3 making this unnecessary.